tests/filecheck/transforms/reuse-memref-space.mlir

```diff
@@ -0,0 +1,319 @@
+// RUN: ./compiler/snax-opt --split-input-file %s -p reuse-memref-space | filecheck %s
+
+builtin.module {
+  func.func @streamer_matmul(%arg0 : memref<?x?xi8, "L3">, %arg1 : memref<?x?xi32, "L3">) {
+    %0 = arith.constant 8 : index
+    %1 = arith.constant 1 : index
+    %2 = arith.constant 0 : i32
+    %3 = arith.constant 0 : index
+    scf.for %arg2 = %3 to %0 step %1{
+        %4 = memref.alloc(%0, %0) {"alignment" = 64 : i64} : memref<?x?xi8, #tsl.tsl<[?, 8] -> (?, 8), [?, 8] -> (256, 1)>, "L1">
+        }
+    func.return
+  }
+}
+// Check if the memref.alloc is moved out of the loop with all operands outside loop
+//CHECK: builtin.module {
+//CHECK-NEXT:   func.func @streamer_matmul(%arg0 : memref<?x?xi8, "L3">, %arg1 : memref<?x?xi32, "L3">) {
+//CHECK-NEXT:     %0 = arith.constant 8 : index
+//CHECK-NEXT:     %1 = arith.constant 1 : index
+//CHECK-NEXT:     %2 = arith.constant 0 : i32
+//CHECK-NEXT:     %3 = arith.constant 0 : index
+//CHECK-NEXT:     %4 = memref.alloc(%0, %0) {"alignment" = 64 : i64} : memref<?x?xi8, #tsl.tsl<[?, 8] -> (?, 8), [?, 8] -> (256, 1)>, "L1">
+//CHECK-NEXT:     scf.for %arg2 = %3 to %0 step %1 {
+//CHECK-NEXT:     }
+//CHECK-NEXT:     func.return
+//CHECK-NEXT:   }
+//CHECK-NEXT: }
+// -----
+
+
+builtin.module {
+  func.func @streamer_matmul(%arg0 : memref<?x?xi8, "L3">, %arg1 : memref<?x?xi32, "L3">) {
+    %0 = arith.constant 8 : index
+    %1 = arith.constant 1 : index
+    %2 = arith.constant 0 : i32
+    %3 = arith.constant 0 : index
+    scf.for %arg2 = %3 to %0 step %1{
+        %4 = arith.constant 16 : index
+        %5 = arith.constant 8 : index
+        %6 = memref.alloc(%4, %5) {"alignment" = 64 : i64} : memref<?x?xi8, #tsl.tsl<[?, 8] -> (?, 8), [?, 8] -> (256, 1)>, "L1">
+        }
+    func.return
+  }
+}
+
+// Check if the memref.alloc is moved out of the loop with all operands inside loop
+//CHECK: builtin.module {
+//CHECK-NEXT:   func.func @streamer_matmul(%arg0 : memref<?x?xi8, "L3">, %arg1 : memref<?x?xi32, "L3">) {
+//CHECK-NEXT:     %0 = arith.constant 8 : index
+//CHECK-NEXT:     %1 = arith.constant 1 : index
+//CHECK-NEXT:     %2 = arith.constant 0 : i32
+//CHECK-NEXT:     %3 = arith.constant 0 : index
+//CHECK-NEXT:     %4 = arith.constant 16 : index
+//CHECK-NEXT:     %5 = arith.constant 8 : index
+//CHECK-NEXT:     %6 = memref.alloc(%4, %5) {"alignment" = 64 : i64} : memref<?x?xi8, #tsl.tsl<[?, 8] -> (?, 8), [?, 8] -> (256, 1)>, "L1">
+//CHECK-NEXT:     scf.for %arg2 = %3 to %0 step %1 {
+//CHECK-NEXT:       %7 = arith.constant 16 : index
+//CHECK-NEXT:       %8 = arith.constant 8 : index
+//CHECK-NEXT:     }
```

Comment:
Instead of creating new operations, I think it makes more sense to hoist the constants outside of the loop too?